crypto/fipsmodule/rand/new_rand.c

```diff
@@ -0,0 +1,290 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <openssl/rand.h>
+#include <openssl/mem.h>
+#include <openssl/ctrdrbg.h>
+#include <openssl/type_check.h>
+
+#include "new_rand_internal.h"
+#include "internal.h"
+#include "../../internal.h"
+
+#include "new_rand_prefix.h"
+#include "entropy/internal.h"
+
+// rand_thread_state contains the per-thread state for the RNG.
+struct rand_thread_local_state {
+  // Thread-local CTR-DRBG state. UBE volatile memory.
+  CTR_DRBG_STATE drbg;
+
+  // generate_calls_since_seed is the number of generate calls made on |drbg|
+  // since it was last (re)seeded. Must be bounded by |kReseedInterval|.
+  uint64_t generate_calls_since_seed;
+
+  // Entropy source. UBE volatile memory.
+  struct entropy_source entropy_source;
+};
+
+// rand_thread_local_state frees a |rand_thread_local_state|. This is called
+// when a thread exits.
+static void rand_thread_local_state_free(void *state_in) {
+
+  struct rand_thread_local_state *state = state_in;
+  if (state_in == NULL) {
+    return;
+  }
+
+  OPENSSL_free(state);
+}
+
+// TODO
+// Ensure snapsafe is in valid state
+static int rand_ensure_valid_state(void) {
+  return 1;
+}
+
+// TODO
+// For UBE.
+static int rand_ensure_ctr_drbg_uniquness(struct rand_thread_local_state *state,
+  size_t out_len) {
+
+  return 1;
+}
+
+// rand_maybe_get_ctr_drbg_pred_resistance maybe fills |pred_resistance| with
+// |RAND_PRED_RESISTANCE_LEN| bytes. The bytes are sourced from the prediction
+// resistance source from the entropy source in |state|, if such a source has
+// been configured.
+//
+// |*pred_resistance_len| is set to 0 if no prediction resistance source is
+// available and |RAND_PRED_RESISTANCE_LEN| otherwise.
+static void rand_maybe_get_ctr_drbg_pred_resistance(
+  struct rand_thread_local_state *state,
+  uint8_t pred_resistance[RAND_PRED_RESISTANCE_LEN],
+  size_t *pred_resistance_len) {
+
+  *pred_resistance_len = 0;
+
+  if (state->entropy_source.prediction_resistance != NULL) {
+    state->entropy_source.prediction_resistance(pred_resistance);
+    *pred_resistance_len = RAND_PRED_RESISTANCE_LEN;
+  }
+}
+
+// rand_get_ctr_drbg_seed_entropy source entropy for seeding and reseeding the
+// CTR-DRBG state. Firstly, |seed| is filled with |CTR_DRBG_ENTROPY_LEN| bytes
+// from the seed source from the |state| entropy source. Secondly, if
+// available, |CTR_DRBG_ENTROPY_LEN| bytes is filled into
+// |personalization_string| sourced from the personalization string source from
+// the entropy source in |state|.
+//
+// |*personalization_string_len| is set to 0 if no personalization string source
+// is available and |CTR_DRBG_ENTROPY_LEN| otherwise.
+static void rand_get_ctr_drbg_seed_entropy(struct entropy_source *entropy_source,
```

Comment:
```suggestion
static void rand_get_ctr_drbg_seed_entropy(struct entropy_source *state,
```

To match your doc string.

Reply:
Changed description instead.